packages/services/src/event/index.ts

```diff
@@ -0,0 +1,237 @@
+   */
+  constructor(options: EventManager.IOptions = {}) {
+    this.serverSettings =
+      options.serverSettings ?? ServerConnection.makeSettings();
+
+    this._poll = new Poll({ factory: () => this._subscribe() });
+    this._stream = new Private.Stream(this);
+
+    // TODO: Remove this logic in JupyterLab 4
+    if (this._isDisabled) {
+      this._stream.stop();
+    } else {
+      void this._poll.start();
+    }
+  }
+
+  /**
+   * The server settings used to make API requests.
+   */
+  readonly serverSettings: ServerConnection.ISettings;
+
+  /**
+   * Whether the event manager is disposed.
+   */
+  get isDisposed(): boolean {
+    return this._isDisposed;
+  }
+
+  /**
+   * An event stream that emits and yields each new event.
+   */
+  get stream(): Event.Stream {
+    return this._stream;
+  }
+
+  /**
+   * Dispose the event manager.
+   */
+  dispose(): void {
+    if (this.isDisposed) {
+      return;
+    }
+    this._isDisposed = true;
+
+    // Clean up poll.
+    this._poll.dispose();
+
+    // Clean up socket.
+    const socket = this._socket;
+    if (socket) {
+      this._socket = null;
+      socket.onopen = () => undefined;
+      socket.onerror = () => undefined;
+      socket.onmessage = () => undefined;
+      socket.onclose = () => undefined;
+      socket.close();
+    }
+
+    // Clean up stream.
+    Signal.clearData(this);
+    this._stream.stop();
+  }
+
+  /**
+   * Post an event request to be emitted by the event bus.
+   */
+  async emit(event: Event.Request): Promise<void> {
+    const { serverSettings } = this;
+    const { baseUrl, token } = serverSettings;
+    const { makeRequest, ResponseError } = ServerConnection;
+    const url =
+      URLExt.join(baseUrl, SERVICE_EVENTS_URL) +
+      (token ? `?token=${token}` : '');
+    const init = { body: JSON.stringify(event), method: 'POST' };
+    const response = await makeRequest(url, init, serverSettings);
+
+    if (response.status !== 204) {
+      throw new ResponseError(response);
+    }
+  }
+
+  /**
+   * Subscribe to event bus emissions.
+   */
+  private _subscribe(): Promise<void> {
+    return new Promise<void>((_, reject) => {
+      if (this.isDisposed || this._isDisabled) {
+        return;
+      }
+
+      const { token, WebSocket, wsUrl } = this.serverSettings;
+      const url =
+        URLExt.join(wsUrl, SERVICE_EVENTS_URL, 'subscribe') +
+        (token ? `?token=${encodeURIComponent(token)}` : '');
+      const socket = (this._socket = new WebSocket(url));
+      const stream = this._stream;
+
+      // Cause the poll to tick a rejection and back off if the socket closes.
+      socket.onclose = () => reject(new Error('EventManager socket closed'));
+      socket.onmessage = msg => msg.data && stream.emit(JSON.parse(msg.data));
+    });
+  }
+
+  // TODO: Remove this check for the `jupyter_server` version.
+  // It is only necessary in JupyterLab < 4.
+  private _isDisabled = 2 > PageConfig.getNotebookVersion()[0];
```

Comment:
Wondering whether that would work in a non-browser environment or in other frontends, since `@jupyterlab/services` can be reused elsewhere?

Reply:
We already use `PageConfig` in `@jupyterlab/services` in a similar way in:

https://github.com/jupyterlab/jupyterlab/blob/9ee65ca27712b19cb8a7a2f4e99453ef1439a525/packages/services/src/serverconnection.ts#L222-L223

I don't think this PR's usage changes the requirements we already imposed in a material way. What do you think, @jtpio?

Reply:
Ah right thanks for the extra context!